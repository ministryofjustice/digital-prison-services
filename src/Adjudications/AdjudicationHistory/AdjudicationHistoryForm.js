import React from 'react'
import PropTypes from 'prop-types'
import GridRow from '@govuk-react/grid-row'
import GridCol from '@govuk-react/grid-col'
import Button from '@govuk-react/button'
import moment from 'moment'
import ErrorSummary from '@govuk-react/error-summary'
import ErrorText from '@govuk-react/error-text'
import { connect } from 'react-redux'
import { Form } from 'react-final-form'
import FormDatePicker from '../../DatePickers/FormDatePicker'
import { linkOnClick } from '../../utils'
import validateThenSubmit from '../../Components/FilterForm/FilterFormValidation'
import {
  FullWidthSelect,
  SearchArea,
  FiltersLabel,
  DateRangeLabel,
  LargeScreenOnlyGridCol,
} from '../../Components/FilterForm/FilterForm.styles'
import { FieldWithError, onHandleErrorClick } from '../../final-form-govuk-helpers'

export const AdjudicationHistoryForm = ({ agencies, search, fieldValues, reset, now }) => (
  <Form
    initialValues={fieldValues}
    onSubmit={values => {
      const errors = validateThenSubmit({
        fromDate: values.fromDate,
        toDate: values.toDate,
        submit: search,
      })(values)
      if (errors) window.scrollTo(0, 0)
      return errors
    }}
    render={({ error, handleSubmit, submitError, submitting, form, values }) => (
      <>
        {error && <ErrorText> {error} </ErrorText>}
        {submitError && (
          <ErrorSummary onHandleErrorClick={onHandleErrorClick} heading="There is a problem" errors={submitError} />
        )}
        <form onSubmit={handleSubmit}>
          <SearchArea>
            <FormFields
              now={now}
              errors={submitError}
              submitting={submitting}
              agencies={agencies}
              values={values}
              reset={() => {
                reset().then(() => form.reset({}))
              }}
            />
          </SearchArea>
        </form>
      </>
    )}
  />
)

const shouldShowDay = now => date => date.isSameOrBefore(now, 'day')

export const FormFields = ({ now, errors, submitting, agencies = [], values, reset }) => (
  <>
    <GridRow>
      <GridCol setDesktopWidth="one-third">
        <GridRow>
          <FiltersLabel>Filters</FiltersLabel>
        </GridRow>
        <GridRow>
          <LargeScreenOnlyGridCol>
            <FieldWithError
              errors={errors}
              name="establishment"
              component={FullWidthSelect}
              label="Establishment"
              id="establishment-select"
              value={values.establishment}
            >
              <option value="">All</option>
              {agencies.map(agency => (
                <option key={agency.agencyId} value={agency.agencyId}>
                  {agency.description}
                </option>
              ))}
            </FieldWithError>
          </LargeScreenOnlyGridCol>
        </GridRow>
      </GridCol>
      <GridCol setDesktopWidth="two-thirds">
        <GridRow>
          <DateRangeLabel>Reported date</DateRangeLabel>
        </GridRow>
        <GridRow>
          <GridCol id="from-date" setDesktopWidth="one-third">
            <FieldWithError
              errors={errors}
              name="fromDate"
              component={FormDatePicker}
              label="From"
              value={values.fromDate}
              shouldShowDay={shouldShowDay(now)}
            />
          </GridCol>
          <GridCol id="to-date" setDesktopWidth="one-third">
            <FieldWithError
              errors={errors}
              name="toDate"
              component={FormDatePicker}
              label="To"
              value={values.toDate}
              shouldShowDay={shouldShowDay(now)}
            />
          </GridCol>
          <GridCol setDesktopWidth="one-third">
            <p className="form-label visible-md visible-lg" />
            <Button id="apply-filter" disabled={submitting} type="submit">
              Apply filters
            </Button>
          </GridCol>
        </GridRow>
      </GridCol>
    </GridRow>
    <a id="clear-filters" className="clear-filters link clickable" {...linkOnClick(reset)}>
      Clear filters
    </a>
  </>
)

AdjudicationHistoryForm.propTypes = {
  now: PropTypes.instanceOf(moment).isRequired,
  search: PropTypes.func.isRequired,
  reset: PropTypes.func.isRequired,
  // mapStateToProps
  agencies: PropTypes.arrayOf(
    PropTypes.shape({
      agencyId: PropTypes.string.isRequired,
      agencyType: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
    }).isRequired
  ).isRequired,
  fieldValues: PropTypes.shape({
    establishment: PropTypes.string,
    fromDate: PropTypes.instanceOf(moment),
    toDate: PropTypes.instanceOf(moment),
  }).isRequired,
}

FormFields.propTypes = {
  now: PropTypes.instanceOf(moment).isRequired,
  errors: PropTypes.arrayOf(PropTypes.shape({})),
  submitting: PropTypes.bool.isRequired,
  agencies: PropTypes.arrayOf(
    PropTypes.shape({
      agencyId: PropTypes.string.isRequired,
      agencyType: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
    }).isRequired
  ).isRequired,
  values: PropTypes.shape({
    establishment: PropTypes.string,
    fromDate: PropTypes.instanceOf(moment),
    toDate: PropTypes.instanceOf(moment),
  }).isRequired,
  reset: PropTypes.func.isRequired,
}

FormFields.defaultProps = {
  errors: [],
}

const mapStateToProps = state => ({
  now: state.adjudicationHistory.now,
  agencies: state.adjudicationHistory.agencies,
  fieldValues: {
    establishment: state.adjudicationHistory.establishment,
    fromDate: state.adjudicationHistory.fromDate,
    toDate: state.adjudicationHistory.toDate,
  },
})

export default connect(mapStateToProps)(AdjudicationHistoryForm)
