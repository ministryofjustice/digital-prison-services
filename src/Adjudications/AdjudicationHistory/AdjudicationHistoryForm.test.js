import React from 'react'
import moment from 'moment'
import testRenderer from 'react-test-renderer'

import AdjudicationHistoryForm from './AdjudicationHistoryForm'

const initialState = {
  adjudicationHistory: {
    now: moment('2017-03-21'),
    pageNumber: 0,
    pageSize: 20,
    totalRecords: 100,
    agencies: [
      { agencyId: 'MDI', agencyType: 'INST', description: 'Moorland' },
      { agencyId: 'LEI', agencyType: 'INST', description: 'Leicester' },
    ],
    establishment: 'MDI',
    fromDate: moment('2017-03-29'),
    toDate: moment('2017-03-29'),
  },
}

describe('Adjudication History Form', () => {
  const store = {}
  const history = {}
  beforeEach(() => {
    store.getState = jest.fn()
    store.subscribe = jest.fn()
    store.dispatch = jest.fn()
    history.push = jest.fn()
    history.replace = jest.fn()
    store.getState.mockReturnValue(initialState)
  })

  it('should render correctly', () => {
    const wrapper = testRenderer
      .create(<AdjudicationHistoryForm store={store} search={jest.fn()} reset={jest.fn()} />)
      .toJSON()

    expect(wrapper).toMatchSnapshot()
  })
})
