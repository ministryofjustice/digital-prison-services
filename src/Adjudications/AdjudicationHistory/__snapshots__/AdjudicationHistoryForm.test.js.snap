// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adjudication History Form should render correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="sc-kgoBCf hNqWyG"
  >
    <div
      className="sc-bdVaJa gkOpjT"
    >
      <div
        className="sc-bwzfXH dGugEg"
      >
        <div
          className="sc-bdVaJa gkOpjT"
        >
          <h3
            className="sc-kpOJdX jmiqVd sc-jzJRlG jslWVk"
          >
            Filters
          </h3>
        </div>
        <div
          className="sc-bdVaJa gkOpjT"
        >
          <div
            className="sc-kGXeez sbhYD sc-bwzfXH iXLolC"
          >
            <label
              className="sc-chPdSV bZWJhx sc-jTzLTM chExuY"
              id="establishment-select"
            >
              <span
                className="sc-VigVT bXITLn"
              >
                Establishment
              </span>
              <select
                className="sc-kAzzGY fdeKIf"
                name="establishment"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                value="MDI"
              >
                <option
                  value=""
                >
                  All
                </option>
                <option
                  value="MDI"
                >
                  Moorland
                </option>
                <option
                  value="LEI"
                >
                  Leicester
                </option>
              </select>
            </label>
          </div>
        </div>
      </div>
      <div
        className="sc-bwzfXH bfdhGR"
      >
        <div
          className="sc-bdVaJa gkOpjT"
        >
          <span
            className="sc-dxgOiQ ihVthS"
          >
            Reported date
          </span>
        </div>
        <div
          className="sc-bdVaJa gkOpjT"
        >
          <div
            className="sc-bwzfXH dGugEg"
            id="from-date"
          >
            <div
              className="rdt"
            >
              <div>
                <label
                  className="sc-jTzLTM gzbVjD"
                >
                  <span
                    className="sc-VigVT bXITLn"
                  >
                     
                    From
                     
                  </span>
                  <input
                    className="form-control sc-fjdhpX bdqmyp"
                    name="fromDate"
                    onChange={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    placeholder=""
                    readOnly={true}
                    type="text"
                    value="29/03/2017"
                  />
                </label>
              </div>
              <div
                className="rdtPicker"
              >
                <div
                  className="rdtDays"
                >
                  <table>
                    <thead>
                      <tr>
                        <th
                          className="rdtPrev"
                          onClick={[Function]}
                        >
                          <span>
                            ‹
                          </span>
                        </th>
                        <th
                          className="rdtSwitch"
                          colSpan={5}
                          data-value={2}
                          onClick={[Function]}
                        >
                          March 2017
                        </th>
                        <th
                          className="rdtNext"
                          onClick={[Function]}
                        >
                          <span>
                            ›
                          </span>
                        </th>
                      </tr>
                      <tr>
                        <th
                          className="dow"
                        >
                          Mo
                        </th>
                        <th
                          className="dow"
                        >
                          Tu
                        </th>
                        <th
                          className="dow"
                        >
                          We
                        </th>
                        <th
                          className="dow"
                        >
                          Th
                        </th>
                        <th
                          className="dow"
                        >
                          Fr
                        </th>
                        <th
                          className="dow"
                        >
                          Sa
                        </th>
                        <th
                          className="dow"
                        >
                          Su
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr>
                        <td
                          className="rdtDay rdtOld"
                          data-value={27}
                          onClick={[Function]}
                        >
                          27
                        </td>
                        <td
                          className="rdtDay rdtOld"
                          data-value={28}
                          onClick={[Function]}
                        >
                          28
                        </td>
                        <td
                          className="rdtDay"
                          data-value={1}
                          onClick={[Function]}
                        >
                          1
                        </td>
                        <td
                          className="rdtDay"
                          data-value={2}
                          onClick={[Function]}
                        >
                          2
                        </td>
                        <td
                          className="rdtDay"
                          data-value={3}
                          onClick={[Function]}
                        >
                          3
                        </td>
                        <td
                          className="rdtDay"
                          data-value={4}
                          onClick={[Function]}
                        >
                          4
                        </td>
                        <td
                          className="rdtDay"
                          data-value={5}
                          onClick={[Function]}
                        >
                          5
                        </td>
                      </tr>
                      <tr>
                        <td
                          className="rdtDay"
                          data-value={6}
                          onClick={[Function]}
                        >
                          6
                        </td>
                        <td
                          className="rdtDay"
                          data-value={7}
                          onClick={[Function]}
                        >
                          7
                        </td>
                        <td
                          className="rdtDay"
                          data-value={8}
                          onClick={[Function]}
                        >
                          8
                        </td>
                        <td
                          className="rdtDay"
                          data-value={9}
                          onClick={[Function]}
                        >
                          9
                        </td>
                        <td
                          className="rdtDay"
                          data-value={10}
                          onClick={[Function]}
                        >
                          10
                        </td>
                        <td
                          className="rdtDay"
                          data-value={11}
                          onClick={[Function]}
                        >
                          11
                        </td>
                        <td
                          className="rdtDay"
                          data-value={12}
                          onClick={[Function]}
                        >
                          12
                        </td>
                      </tr>
                      <tr>
                        <td
                          className="rdtDay"
                          data-value={13}
                          onClick={[Function]}
                        >
                          13
                        </td>
                        <td
                          className="rdtDay"
                          data-value={14}
                          onClick={[Function]}
                        >
                          14
                        </td>
                        <td
                          className="rdtDay"
                          data-value={15}
                          onClick={[Function]}
                        >
                          15
                        </td>
                        <td
                          className="rdtDay"
                          data-value={16}
                          onClick={[Function]}
                        >
                          16
                        </td>
                        <td
                          className="rdtDay"
                          data-value={17}
                          onClick={[Function]}
                        >
                          17
                        </td>
                        <td
                          className="rdtDay"
                          data-value={18}
                          onClick={[Function]}
                        >
                          18
                        </td>
                        <td
                          className="rdtDay"
                          data-value={19}
                          onClick={[Function]}
                        >
                          19
                        </td>
                      </tr>
                      <tr>
                        <td
                          className="rdtDay"
                          data-value={20}
                          onClick={[Function]}
                        >
                          20
                        </td>
                        <td
                          className="rdtDay"
                          data-value={21}
                          onClick={[Function]}
                        >
                          21
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={22}
                        >
                          22
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={23}
                        >
                          23
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={24}
                        >
                          24
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={25}
                        >
                          25
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={26}
                        >
                          26
                        </td>
                      </tr>
                      <tr>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={27}
                        >
                          27
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={28}
                        >
                          28
                        </td>
                        <td
                          className="rdtDay rdtActive rdtDisabled"
                          data-value={29}
                        >
                          29
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={30}
                        >
                          30
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={31}
                        >
                          31
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={1}
                        >
                          1
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={2}
                        >
                          2
                        </td>
                      </tr>
                      <tr>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={3}
                        >
                          3
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={4}
                        >
                          4
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={5}
                        >
                          5
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={6}
                        >
                          6
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={7}
                        >
                          7
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={8}
                        >
                          8
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={9}
                        >
                          9
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
          <div
            className="sc-bwzfXH dGugEg"
            id="to-date"
          >
            <div
              className="rdt"
            >
              <div>
                <label
                  className="sc-jTzLTM gzbVjD"
                >
                  <span
                    className="sc-VigVT bXITLn"
                  >
                     
                    To
                     
                  </span>
                  <input
                    className="form-control sc-fjdhpX bdqmyp"
                    name="toDate"
                    onChange={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    placeholder=""
                    readOnly={true}
                    type="text"
                    value="29/03/2017"
                  />
                </label>
              </div>
              <div
                className="rdtPicker"
              >
                <div
                  className="rdtDays"
                >
                  <table>
                    <thead>
                      <tr>
                        <th
                          className="rdtPrev"
                          onClick={[Function]}
                        >
                          <span>
                            ‹
                          </span>
                        </th>
                        <th
                          className="rdtSwitch"
                          colSpan={5}
                          data-value={2}
                          onClick={[Function]}
                        >
                          March 2017
                        </th>
                        <th
                          className="rdtNext"
                          onClick={[Function]}
                        >
                          <span>
                            ›
                          </span>
                        </th>
                      </tr>
                      <tr>
                        <th
                          className="dow"
                        >
                          Mo
                        </th>
                        <th
                          className="dow"
                        >
                          Tu
                        </th>
                        <th
                          className="dow"
                        >
                          We
                        </th>
                        <th
                          className="dow"
                        >
                          Th
                        </th>
                        <th
                          className="dow"
                        >
                          Fr
                        </th>
                        <th
                          className="dow"
                        >
                          Sa
                        </th>
                        <th
                          className="dow"
                        >
                          Su
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr>
                        <td
                          className="rdtDay rdtOld"
                          data-value={27}
                          onClick={[Function]}
                        >
                          27
                        </td>
                        <td
                          className="rdtDay rdtOld"
                          data-value={28}
                          onClick={[Function]}
                        >
                          28
                        </td>
                        <td
                          className="rdtDay"
                          data-value={1}
                          onClick={[Function]}
                        >
                          1
                        </td>
                        <td
                          className="rdtDay"
                          data-value={2}
                          onClick={[Function]}
                        >
                          2
                        </td>
                        <td
                          className="rdtDay"
                          data-value={3}
                          onClick={[Function]}
                        >
                          3
                        </td>
                        <td
                          className="rdtDay"
                          data-value={4}
                          onClick={[Function]}
                        >
                          4
                        </td>
                        <td
                          className="rdtDay"
                          data-value={5}
                          onClick={[Function]}
                        >
                          5
                        </td>
                      </tr>
                      <tr>
                        <td
                          className="rdtDay"
                          data-value={6}
                          onClick={[Function]}
                        >
                          6
                        </td>
                        <td
                          className="rdtDay"
                          data-value={7}
                          onClick={[Function]}
                        >
                          7
                        </td>
                        <td
                          className="rdtDay"
                          data-value={8}
                          onClick={[Function]}
                        >
                          8
                        </td>
                        <td
                          className="rdtDay"
                          data-value={9}
                          onClick={[Function]}
                        >
                          9
                        </td>
                        <td
                          className="rdtDay"
                          data-value={10}
                          onClick={[Function]}
                        >
                          10
                        </td>
                        <td
                          className="rdtDay"
                          data-value={11}
                          onClick={[Function]}
                        >
                          11
                        </td>
                        <td
                          className="rdtDay"
                          data-value={12}
                          onClick={[Function]}
                        >
                          12
                        </td>
                      </tr>
                      <tr>
                        <td
                          className="rdtDay"
                          data-value={13}
                          onClick={[Function]}
                        >
                          13
                        </td>
                        <td
                          className="rdtDay"
                          data-value={14}
                          onClick={[Function]}
                        >
                          14
                        </td>
                        <td
                          className="rdtDay"
                          data-value={15}
                          onClick={[Function]}
                        >
                          15
                        </td>
                        <td
                          className="rdtDay"
                          data-value={16}
                          onClick={[Function]}
                        >
                          16
                        </td>
                        <td
                          className="rdtDay"
                          data-value={17}
                          onClick={[Function]}
                        >
                          17
                        </td>
                        <td
                          className="rdtDay"
                          data-value={18}
                          onClick={[Function]}
                        >
                          18
                        </td>
                        <td
                          className="rdtDay"
                          data-value={19}
                          onClick={[Function]}
                        >
                          19
                        </td>
                      </tr>
                      <tr>
                        <td
                          className="rdtDay"
                          data-value={20}
                          onClick={[Function]}
                        >
                          20
                        </td>
                        <td
                          className="rdtDay"
                          data-value={21}
                          onClick={[Function]}
                        >
                          21
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={22}
                        >
                          22
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={23}
                        >
                          23
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={24}
                        >
                          24
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={25}
                        >
                          25
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={26}
                        >
                          26
                        </td>
                      </tr>
                      <tr>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={27}
                        >
                          27
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={28}
                        >
                          28
                        </td>
                        <td
                          className="rdtDay rdtActive rdtDisabled"
                          data-value={29}
                        >
                          29
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={30}
                        >
                          30
                        </td>
                        <td
                          className="rdtDay rdtDisabled"
                          data-value={31}
                        >
                          31
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={1}
                        >
                          1
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={2}
                        >
                          2
                        </td>
                      </tr>
                      <tr>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={3}
                        >
                          3
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={4}
                        >
                          4
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={5}
                        >
                          5
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={6}
                        >
                          6
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={7}
                        >
                          7
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={8}
                        >
                          8
                        </td>
                        <td
                          className="rdtDay rdtNew rdtDisabled"
                          data-value={9}
                        >
                          9
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
          <div
            className="sc-bwzfXH dGugEg"
          >
            <p
              className="form-label visible-md visible-lg"
            />
            <button
              className="sc-htpNat hfQKkJ"
              disabled={false}
              id="apply-filter"
              type="submit"
            >
              Apply filters
            </button>
          </div>
        </div>
      </div>
    </div>
    <a
      className="clear-filters link clickable"
      id="clear-filters"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="link"
      tabIndex={0}
    >
      Clear filters
    </a>
  </div>
</form>
`;
